test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs

```diff
@@ -136,5 +141,48 @@ private byte[] FormatMessageToArray(byte[] message)
                 Closed?.Invoke(ex);
             }
         }
+
+        public IDisposable OnReceived(Func<byte[], object, Task> callback, object state)
+        {
+            var receiveCallBack = new ReceiveCallBack(callback, state);
+            _callbacks.Add(receiveCallBack);
+            return new Subscription(receiveCallBack, _callbacks);
+        }
+
+        private class ReceiveCallBack
+        {
+            public readonly Func<byte[], object, Task> CallBack;
+            private readonly object State;
+
+            public ReceiveCallBack(Func<byte[], object, Task> callBack, object state)
+            {
+                CallBack = callBack;
+                State = state;
+            }
+
+            public Task InvokeAsync(byte[] data)
+            {
+                return CallBack(data, State);
+            }
+        }
+
+        private class Subscription : IDisposable
+        {
+            private readonly ReceiveCallBack _callback;
+            private readonly List<ReceiveCallBack> _callbacks;
+            public Subscription(ReceiveCallBack callback, List<ReceiveCallBack> callbacks)
+            {
+                _callback = callback;
+                _callbacks = callbacks;
+            }
+
+            public void Dispose()
+            {
+                lock (_callbacks)
```

Comment:
locking in just one place does not make much sense. I was thinking we should make the callback list threadsafe but since we don't add/remove callbacks in our tests after the connection is started we should be fine. We might need to do this in the future if we start seeing errors caused by this.